source/drivers/MicroBitFile.cpp

```diff
@@ -0,0 +1,280 @@
+
+    buff[size] = 0;
+
+    int ret = read(buff, size);
+
+    if(ret < 0)
+        return ManagedString();
+
+    return ManagedString(buff,ret);
+}
+
+/**
+  * Removes this MicroBitFile from the MicroBitFileSystem.
+  *
+  * @return MICROBIT_OK on success, MICROBIT_INVALID_PARAMETER if the given filename
+  *         does not exist, MICROBIT_CANCELLED if something went wrong.
+  */
+int MicroBitFile::remove()
+{
+    if(fileHandle < 0)
+        return MICROBIT_NOT_SUPPORTED;
+
+    int ret = MicroBitFileSystem::defaultFileSystem->remove(fileName.toCharArray());
+
+    if(ret < 0)
+        return ret;
+
+    fileHandle = MICROBIT_NOT_SUPPORTED;
+
+    return ret;
+}
+
+/**
+  * Seeks to the end of the file, and appends the given ManagedString to this MicroBitFile instance.
+  *
+  * @param bytes a pointer to the bytes to write to this file.
+  *
+  * @param len the number of bytes to write.
+  *
+  * @return The number of bytes written, MICROBIT_NOT_SUPPORTED if the current file
+  *         handle is invalid or this file was not opened in WRITE mode.
+  */
+int MicroBitFile::append(const char *bytes, int len)
+{
+    if(fileHandle < 0)
+        return MICROBIT_NOT_SUPPORTED;
+
+    int ret =  MicroBitFileSystem::defaultFileSystem->seek(fileHandle, 0, MB_SEEK_END);
+
+    if(ret < 0)
+        return ret;
+
+    return write(bytes,len);
+}
+
+/**
+  * Seeks to the end of the file, and appends the given ManagedString to this MicroBitFile instance.
+  *
+  * @param s The ManagedString to write to this file.
+  *
+  * @return The number of bytes written, MICROBIT_NOT_SUPPORTED if the current file
+  *         handle is invalid or this file was not opened in WRITE mode.
+  */
+int MicroBitFile::append(ManagedString s)
+{
+    return append(s.toCharArray(), s.length());
+}
+
+/**
+  * Returns the handle used by this MicroBitFile instance.
+  *
+  * @note This member function will also inform the user of any errors encountered
+  *       during the opening of this MicroBitFile. At open, the handle is set
+  *       to the return value from MicroBitFileSystem.open().
+  */
+int MicroBitFile::getHandle()
+{
+    return fileHandle;
+}
+
+/**
+  * Opens this MicroBitFile instance if the file has previously been closed.
+  *
+  * @param mode One of: READ, WRITE, READ_AND_WRITE. Defaults to READ_AND_WRITE.
+  *
+  * @return MICROBIT_OK on success, MICROBIT_NOT_SUPPORTED if the file is already open,
+  *         MICROBIT_INVALID_PARAMETER if the filename is too large, MICROBIT_NO_RESOURCES
+  *         if the file system is full, or memory could not be allocated..
+  *
+  * @note MicroBitFiles are opened at construction and are implicitly closed at
+  *       destruction. They can be closed explicitly using the close() member function.
+  */
+int MicroBitFile::open(int mode)
+{
+    if(fileHandle >= 0)
+        return MICROBIT_NOT_SUPPORTED;
+
+    int ret = MicroBitFileSystem::defaultFileSystem->open(fileName.toCharArray(), mode | CREATE);
```

Comment:
Can you give a use case for open().  As far as I can see it's not possible to create the MicrobitFile object with having opened it (which would involve creating it if necessary).   Once the object is created then the only reason for calling open() would be if I've called close() on it (and why would I do that, rather than create a new object?), and if I've called close() then the underlying file still exists, and so I don't need the CREATE flag.  The only time the underlying file doesn't exist is if I call remove() - and if I've done that adding the CREATE flag prevents me using open() as a test of existence ... I guess my real question is why the same arguments that we used for not doing this on the constructor don't apply here?

Couldn't we get the same default behaviour by doing the same as we did for the constructor by making the default for mode = READ_AND_WRITE | CREATE

Maybe I'm missing something ...